### PR TITLE
Llama: make slow tests green 🟢 

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Part 1 of #32685 -- update our tests to be sure we don't break things 🤗 

Makes slow `llama` tests happy on my local environment. Some tests are still failing on our slow CI, mostly due to hardware (e.g. out of memory), but they do not have an impact on #32685 [in other words, in checking the correctness of new changes].

The exception is `test_compile_static_cache`, which passes when run in isolation but fails if run with other tests due to `accelerate` + `torch.compile` incompatibilities (see our internal discussion [here](https://huggingface.slack.com/archives/C01Q6JPP6NA/p1724694563976419)).

### CODE REVIEW

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -243,30 +245,33 @@ def _ignore_causal_mask_sdpa(
         is_training: bool = False,
     ) -> bool:
         """
-        Detects whether the optional user-specified attention_mask & the automatically created causal mask can be ignored in case PyTorch's SDPA is used, rather relying on SDPA's `is_causal` argument.
+        Detects whether the optional user-specified attention_mask & the automatically created causal mask can be
```
Comment (gante) [inline src/transformers/modeling_attn_mask_utils.py:248]: This function had many >120 char lines, which did not fit on my screen :D The diff in docstrings/comments is exclusively due to breaking likes.

File: src/transformers/modeling_attn_mask_utils.py [hunk 2]
```diff
@@ -243,30 +245,33 @@ def _ignore_causal_mask_sdpa(
...
+        passed).
         """
 
         _, query_length = inputs_embeds.shape[0], inputs_embeds.shape[1]
         key_value_length = query_length + past_key_values_length
 
-        is_tracing = (
-            torch.jit.is_tracing()
-            or isinstance(inputs_embeds, torch.fx.Proxy)
-            or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
```
Comment (gante) [inline src/transformers/modeling_attn_mask_utils.py:259]: `is_torchdynamo_compiling()` is a more comprehensive check.

I've replaced all occurrences of this pattern.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 3]
```diff
@@ -790,11 +790,6 @@ def _update_causal_mask(
         past_key_values: Cache,
         output_attentions: bool,
     ):
-        # TODO: As of torch==2.2.0, the `attention_mask` passed to the model in `generate` is 2D and of dynamic length even when the static
-        # KV cache is used. This is an issue for torch.compile which then recaptures cudagraphs at each decode steps due to the dynamic shapes.
-        # (`recording cudagraph tree for symint key 13`, etc.), which is VERY slow. A workaround is `@torch.compiler.disable`, but this prevents using
-        # `fullgraph=True`. See more context in https://github.com/huggingface/transformers/pull/29114
-
```
Comment (gante) [inline src/transformers/models/bloom/modeling_bloom.py:797]: No longer true with https://github.com/huggingface/transformers/issues/32227

Removed all occurrences of this comment.

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -1815,22 +1817,14 @@ def test_generate_compile_fullgraph(self):
...
-                # eager static cache
-                torch.compiler.reset()
-                model.generation_config.cache_implementation = "static"
-                output_static = model.generate(model_inputs, **generation_kwargs)
-                self.assertListEqual(output_dynamic.tolist(), output_static.tolist())
-
-                # compiled static cache (removes the cache initialized in the previous check, to confirm we can
-                # initialize the cache in full compiled mode)
-                model._cache = None
+                # end-to-end compiled dynamic cache
```
Comment (gante) [inline tests/generation/test_utils.py:1823]: the reference for end-to-end compilation is with dynamic cache -- I forgot to update this test before the last set of changes in the PR that introduced end-to-end compilation :)

File: tests/models/llama/test_modeling_llama.py [hunk 5]
```diff
@@ -779,8 +781,8 @@ def test_model_7b_logits_bf16(self):
             torch.allclose(
                 EXPECTED_SLICE[self.cuda_compute_capability_major_version].to(torch_device),
                 out.logits[0, 0, :15],
-                atol=1e-3,
-                rtol=1e-3,
+                atol=1e-2,
```
Comment (gante) [inline tests/models/llama/test_modeling_llama.py:784]: We have expected results built for the GPUs in our CI, T4 and A10. However, if I navigate to the original commit and install the torch version at the time (torch 2.3), the test fails on my RTX4090

This means there are probably tiny device-related differences whose explanation is beyond the cuda compute capability major version. 

As such, increased the tolerance. To be fair, `1e-2` is within the expected differences for 16-bit computations, just like `1e-5` is for 32-bit computations.

File: src/transformers/modeling_attn_mask_utils.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/modeling_attn_mask_utils.py:259]: is torchdynamo_compiling takes into account the version of torch?

File: tests/generation/test_utils.py (hunk 4 above)
Comment (ArthurZucker) [reply tests/generation/test_utils.py:1823]: this includes recompile no? Or dynamic shapes are handled now?

**Review by ArthurZucker — APPROVED** [review]
thanks

File: tests/generation/test_utils.py (hunk 4 above)
Comment (gante) [reply tests/generation/test_utils.py:1823]: Yes, it issues a recompile

(but static cache doesn't work at the moment -- both things, recompilations and using static caches, need to be addressed)

**Pull request merged**